src/connectivity/johnson_simple_cycles.rs

```diff
@@ -0,0 +1,309 @@
+                // (so no updates to stack) and populate it with an empty
+                // set.
+                None => {
+                    block.insert(stack_node, HashSet::new());
+                }
+            }
+            blocked.remove(&stack_node);
+        }
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn process_stack(
+    start_node: NodeIndex,
+    stack: &mut Vec<(NodeIndex, IndexSet<NodeIndex>)>,
+    path: &mut Vec<NodeIndex>,
+    closed: &mut HashSet<NodeIndex>,
+    blocked: &mut HashSet<NodeIndex>,
+    block: &mut HashMap<NodeIndex, HashSet<NodeIndex>>,
+    subgraph: &StableDiGraph<(), ()>,
+    reverse_node_map: &HashMap<NodeIndex, NodeIndex>,
+) -> Option<IterNextOutput<NodeIndices, &'static str>> {
+    while let Some((this_node, neighbors)) = stack.last_mut() {
+        if let Some(next_node) = neighbors.pop() {
+            if next_node == start_node {
+                // Out path in input graph basis
+                let mut out_path: Vec<usize> = Vec::with_capacity(path.len());
+                for n in path {
+                    out_path.push(reverse_node_map[n].index());
+                    closed.insert(*n);
+                }
+                return Some(IterNextOutput::Yield(NodeIndices { nodes: out_path }));
+            } else if blocked.insert(next_node) {
+                path.push(next_node);
+                stack.push((
+                    next_node,
+                    subgraph
+                        .neighbors(next_node)
+                        .collect::<IndexSet<NodeIndex>>(),
+                ));
+                closed.remove(&next_node);
+                blocked.insert(next_node);
+                continue;
+            }
+        }
+        if neighbors.is_empty() {
+            if closed.contains(this_node) {
+                unblock(*this_node, blocked, block);
+            } else {
+                for neighbor in subgraph.neighbors(*this_node) {
+                    let block_neighbor = block.entry(neighbor).or_insert_with(HashSet::new);
+                    block_neighbor.insert(*this_node);
+                }
+            }
+            stack.pop();
+            path.pop();
+        }
+    }
+    None
+}
+
+#[pymethods]
+impl SimpleCycleIter {
+    fn __iter__(slf: PyRef<Self>) -> Py<SimpleCycleIter> {
+        slf.into()
+    }
+
+    fn __next__(mut slf: PyRefMut<Self>) -> PyResult<IterNextOutput<NodeIndices, &'static str>> {
+        if slf.self_cycles.is_some() {
+            let self_cycles = slf.self_cycles.as_mut().unwrap();
+            let cycle_node = self_cycles.pop().unwrap();
+            if self_cycles.is_empty() {
+                slf.self_cycles = None;
+            }
+            return Ok(IterNextOutput::Yield(NodeIndices {
+                nodes: vec![cycle_node.index()],
+            }));
+        }
+        // Restore previous state if it exists
+        let mut stack: Vec<(NodeIndex, IndexSet<NodeIndex>)> = slf.stack.drain(..).collect();
```

Comment:
[`mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) might be more performant.
```suggestion
        let mut stack: Vec<(NodeIndex, IndexSet<NodeIndex>)> = mem::take(&mut slf.stack);
```

Reply:
oh, that's a good call. I'll update these to use `mem::take()`

Reply:
Done in: https://github.com/Qiskit/rustworkx/pull/633/commits/388111e2d1c5159364b9a4d82a603ec79749a9be